snapshots/lvm/snapshotter.go

```diff
@@ -0,0 +1,416 @@
+		}
+	}
+	if _, err = toggleactivateLV(config.VgName, metavolume, true); err != nil {
+		return nil, errors.Wrap(err, "Unable to activate metavolume")
+	}
+
+	metavolpath := filepath.Join(config.RootPath, metaVolumeMountName)
+	if errdir := os.MkdirAll(metavolpath, 0700); errdir != nil {
+		return nil, errdir
+	}
+
+	metamount := []mount.Mount{
+		{
+			Source:  filepath.Join("/dev", config.VgName, metavolume),
+			Type:    config.FsType,
+			Options: []string{},
+		},
+	}
+
+	if err = mount.All(metamount, metavolpath); err != nil {
+		return nil, err
+	}
+	ms, err := storage.NewMetaStore(filepath.Join(metavolpath, "metadata.db"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &snapshotter{
+		config:      config,
+		ms:          ms,
+		metaVolPath: metavolpath,
+	}, nil
+}
+
+// Stat returns the info for an active or committed snapshot by name or
+// key.
+//
+// Should be used for parent resolution, existence checks and to discern
+// the kind of snapshot.
+func (o *snapshotter) Stat(ctx context.Context, key string) (snapshots.Info, error) {
+	log.G(ctx).Debugf("Stat called for: %s", key)
+	ctx, t, err := o.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return snapshots.Info{}, err
+	}
+	defer func() {
+		if rerr := t.Rollback(); rerr != nil {
+			log.G(ctx).WithError(rerr).Warn("Failed to rollback transaction")
+		}
+	}()
+	_, info, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return snapshots.Info{}, err
+	}
+
+	return info, nil
+}
+
+func (o *snapshotter) Update(ctx context.Context, info snapshots.Info, fieldpaths ...string) (snapshots.Info, error) {
+	log.G(ctx).Debugf("Update called for : %+v", info)
+	ctx, t, err := o.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return snapshots.Info{}, err
+	}
+
+	info, err = storage.UpdateInfo(ctx, info, fieldpaths...)
+	if err != nil {
+		if rerr := t.Rollback(); err != nil {
+			log.G(ctx).WithError(rerr).Warn("Failed to rollback transaction")
+		}
+		return snapshots.Info{}, err
+	}
+
+	if err := t.Commit(); err != nil {
+		return snapshots.Info{}, err
+	}
+
+	return info, nil
+}
+
+func (o *snapshotter) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
+	log.G(ctx).Debugf("Usage of key %+v", key)
+	ctx, t, err := o.ms.TransactionContext(ctx, false)
+	var s storage.Snapshot
+	var du fs.Usage
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	defer func() {
+		if rerr := t.Rollback(); err != nil {
+			log.G(ctx).WithError(rerr).Warn("Failed to rollback transaction")
+		}
+	}()
+	id, info, usage, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+
+	if info.Kind == snapshots.KindActive {
+		if s, err = storage.GetSnapshot(ctx, key); err != nil {
+			return snapshots.Usage{}, err
+		}
+		mounts := o.mounts(s)
+		if err = mount.WithTempMount(ctx, mounts, func(root string) error {
+			if du, err = fs.DiskUsage(ctx, o.getSnapshotDir(id)); err != nil {
+				return err
```

Comment:
Hi, I don't understand why not use  DiskUsage(..., root)? "root" is the tmep dir, right?

Reply:
Yes indeed. The mount logic was added later and i missed this bit. Will fix this one too. Thanks for catching this.